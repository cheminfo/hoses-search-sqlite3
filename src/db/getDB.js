import fs, { existsSync, mkdirSync } from 'fs';
import { join } from 'path';

import sqLite from 'better-sqlite3';

let db;

export default function getDB() {
  const path = new URL('../../sqlite/', import.meta.url).pathname;
  console.log({ path })
  if (!db) {
    if (!existsSync(path)) {
      mkdirSync(path);
    }
    db = sqLite(join(path, 'db.sqlite'));
    // https://www.sqlite.org/wal.html
    // Activating WAL mode allows to get a speed improvement of 100x !!!
    db.pragma('journal_mode = WAL');
  }

  setInterval(() => {
    fs.stat(join(path, 'db.sqlite-wal'), (err, stat) => {
      if (err) {
        if (err.code !== 'ENOENT') throw err;
      } else if (stat.size > 100000000) {
        db.pragma('wal_checkpoint(RESTART)');
      }
    });
  }, 300000).unref();

  const createTable = `
CREATE TABLE IF NOT EXISTS molecules (
  id data_type PRIMARY KEY,
  idCode data_type TEXT,
  ssIndex0 data_type INTEGER,
  ssIndex1 data_type INTEGER,
  ssIndex2 data_type INTEGER,
  ssIndex3 data_type INTEGER,
  ssIndex4 data_type INTEGER,
  ssIndex5 data_type INTEGER,
  ssIndex6 data_type INTEGER,
  ssIndex7 data_type INTEGER
);
CREATE INDEX IF NOT EXISTS ssIndex0 ON molecules(ssIndex0);
CREATE INDEX IF NOT EXISTS ssIndex1 ON molecules(ssIndex1);
CREATE INDEX IF NOT EXISTS ssIndex2 ON molecules(ssIndex2);
CREATE INDEX IF NOT EXISTS ssIndex3 ON molecules(ssIndex3);
CREATE INDEX IF NOT EXISTS ssIndex4 ON molecules(ssIndex4);
CREATE INDEX IF NOT EXISTS ssIndex5 ON molecules(ssIndex5);
CREATE INDEX IF NOT EXISTS ssIndex6 ON molecules(ssIndex6);
CREATE INDEX IF NOT EXISTS ssIndex7 ON molecules(ssIndex7);
`;

  db.exec(createTable);
  return db;
}
