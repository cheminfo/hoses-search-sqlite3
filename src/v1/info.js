import debugLibrary from 'debug';

const debug = debugLibrary('getInfoFromSmiles');

export default function info(fastify) {
  fastify.route({
    url: '/v1/info',
    method: ['GET', 'POST'],
    handler: getInfo,
    scheme: {
      summary: '$retrieve information from a molfile',
      description: '',
    },
  });
}

async function getInfo(request, response) {
  const body = request.query;
  try {
    const result = {};
    return await response.send({ result });
  } catch (error) {
    debug(`Error: ${error.stack}`);
    return response.send({ result: {}, log: error.toString() });
  }
}

export function getAlgorithmsWithContact(db) {
  const infoStmt = db.prepare(`
  SELECT algorithmID, name, version, description, algorithms.contactID, mail
  FROM algorithms
  INNER JOIN contacts ON algorithms.contactID = contacts.contactID`);
  return infoStmt.all();
}

export function getAtomCountPerAlgorithm(db) {
  const AtomCountstmt = db.prepare(`
SELECT energies.algorithmID, algorithms.name, algorithms.version, COUNT(energies.atomID) AS atomCount
FROM algorithms
INNER JOIN energies ON algorithms.algorithmID = energies.algorithmID
GROUP BY energies.algorithmID`);
  return AtomCountstmt.all();
}

export function getHoseCountPerAlgorithm(db) {
  const hoseCountStmt = db.prepare(`
SELECT energies.algorithmID, COUNT(hoseCodes.hoseID) AS hoseCount
FROM atoms
INNER JOIN hoseCodes ON hoseCodes.atomID = atoms.atomID
INNER JOIN energies ON energies.atomID = atoms.atomID
GROUP BY energies.algorithmID`);
  return hoseCountStmt.all();
}

export function getEntryCountPerAlgorithm(db) {
  const EntryCountStmt = db.prepare(`
SELECT energies.algorithmID, COUNT(entries.entryID) AS entryCount
FROM atoms
INNER JOIN entries ON atoms.entryID = entries.entryID
INNER JOIN energies ON atoms.atomID = energies.atomID
GROUP BY energies.algorithmID`);
  return EntryCountStmt.all();
}
