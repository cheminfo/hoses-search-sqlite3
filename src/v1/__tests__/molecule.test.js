import { readFileSync } from 'node:fs';

import { test, expect } from 'vitest';

import { getTempDB } from '../../db/getDB.js';
import { importXYZ } from '../../import/importXYZ.js';
import { getMoleculesFromDB } from '../molecules.js';

import { testDataProperties } from './data/test.info.js';

test('getMolecules', async () => {
  const contents = [
    readFileSync(new URL('data/test_1.xyz', import.meta.url)),
    readFileSync(new URL('data/test_2.xyz', import.meta.url)),
    readFileSync(new URL('data/test_3.xyz', import.meta.url)),
  ];
  const db = await getTempDB();
  for (let i = 0; i < contents.length; i++) {
    await importXYZ(contents[i], db, testDataProperties[i]);
  }

  const moleculeInfo = getMoleculesFromDB('test2@test.ch', 'GW_charged', db);
  //   console.log(moleculeInfo);
});
