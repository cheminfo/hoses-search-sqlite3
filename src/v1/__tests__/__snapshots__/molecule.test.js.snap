// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getMolecules > molecules 1`] = `
[
  {
    "atoms": [
      {
        "hoses": [
          {
            "idCode": "eF@Hp\\QPZQgr]W~d\`xUP",
            "sphere": 1,
          },
        ],
        "label": undefined,
        "value": 290.2325,
      },
      {
        "hoses": [
          {
            "idCode": "fH@NJ\`uOkoth\\Jh",
            "sphere": 0,
          },
        ],
        "label": "C",
        "value": 290.2325,
      },
      {
        "hoses": [
          {
            "idCode": "gC\`@DfjdPx\`hMDLTy|Qtm[nRHGBj@",
            "sphere": 2,
          },
        ],
        "label": "C",
        "value": 290.2325,
      },
      {
        "hoses": [
          {
            "idCode": "gJQ@@dmUQDCbZ\`uPqSdojKfj}q{}Hf\\Jh",
            "sphere": 3,
          },
        ],
        "label": "C",
        "value": 290.2325,
      },
      {
        "hoses": [
          {
            "idCode": "gJQ@@dmUQDCbZ\`uPqSdojKfj}q{}Hf\\Jh",
            "sphere": 4,
          },
        ],
        "label": "C",
        "value": 290.2325,
      },
      {
        "hoses": [
          {
            "idCode": "fH@NJ\`uOjth\\Jh",
            "sphere": 0,
          },
        ],
        "label": "C",
        "value": 290.6963,
      },
      {
        "hoses": [
          {
            "idCode": "gC\`@H}PGDEAhabgObKfkmwzQ@xUP",
            "sphere": 1,
          },
        ],
        "label": "C",
        "value": 290.6963,
      },
      {
        "hoses": [
          {
            "idCode": "gJQ@@dmU@\\QTFjFJ\\e}Q\\uWnO_iDSaU@",
            "sphere": 2,
          },
        ],
        "label": "C",
        "value": 290.6963,
      },
      {
        "hoses": [
          {
            "idCode": "gJQ@@dmU@\\QTFjFJ\\e}Q\\uWnO_iDSaU@",
            "sphere": 3,
          },
        ],
        "label": "C",
        "value": 290.6963,
      },
      {
        "hoses": [
          {
            "idCode": "gJQ@@dmU@\\QTFjFJ\\e}Q\\uWnO_iDSaU@",
            "sphere": 4,
          },
        ],
        "label": "C",
        "value": 290.6963,
      },
      {
        "hoses": [
          {
            "idCode": "eF@Hp\\QPZQgr]W~d\`xUP",
            "sphere": 1,
          },
        ],
        "label": "C",
        "value": 290.4741,
      },
      {
        "hoses": [
          {
            "idCode": "fH@NJ\`uOkoth\\Jh",
            "sphere": 0,
          },
        ],
        "label": "C",
        "value": 290.4741,
      },
      {
        "hoses": [
          {
            "idCode": "gC\`@Dfjd\`x\`hMDLTy|Qtm[nRHGBj@",
            "sphere": 2,
          },
        ],
        "label": "C",
        "value": 290.4741,
      },
      {
        "hoses": [
          {
            "idCode": "gJQ@@dmUQDCbZ\`uPqSdojKfj}q{}Hf\\Jh",
            "sphere": 3,
          },
        ],
        "label": "C",
        "value": 290.4741,
      },
      {
        "hoses": [
          {
            "idCode": "gJQ@@dmUQDCbZ\`uPqSdojKfj}q{}Hf\\Jh",
            "sphere": 4,
          },
        ],
        "label": "C",
        "value": 290.4741,
      },
      {
        "hoses": [
          {
            "idCode": "eMHAIhNJhMXrolnuo}IIpj\`",
            "sphere": 1,
          },
        ],
        "label": "C",
        "value": 292.0451,
      },
      {
        "hoses": [
          {
            "idCode": "fH@NJ\`uOk_th\\Jh",
            "sphere": 0,
          },
        ],
        "label": "C",
        "value": 292.0451,
      },
      {
        "hoses": [
          {
            "idCode": "gJQ@@dmU@\\RTFjFJ\\e}Q\\uWnO_iDcaU@",
            "sphere": 2,
          },
        ],
        "label": "C",
        "value": 292.0451,
      },
      {
        "hoses": [
          {
            "idCode": "gJQ@@dmU@\\RTFjFJ\\e}Q\\uWnO_iDcaU@",
            "sphere": 3,
          },
        ],
        "label": "C",
        "value": 292.0451,
      },
      {
        "hoses": [
          {
            "idCode": "gJQ@@dmU@\\RTFjFJ\\e}Q\\uWnO_iDcaU@",
            "sphere": 4,
          },
        ],
        "label": "C",
        "value": 292.0451,
      },
      {
        "hoses": [
          {
            "idCode": "eFHBLGDTFdY|eviHOeT",
            "sphere": 1,
          },
        ],
        "label": "O",
        "value": 538.0942,
      },
      {
        "hoses": [
          {
            "idCode": "eMHAIhNHhMXrolnuo}IA|j\`",
            "sphere": 2,
          },
        ],
        "label": "O",
        "value": 538.0942,
      },
      {
        "hoses": [
          {
            "idCode": "fI@GEPZgu_zTOeT",
            "sphere": 0,
          },
        ],
        "label": "O",
        "value": 538.0942,
      },
      {
        "hoses": [
          {
            "idCode": "gJQ@@dmU@\\PTFjFJ\\e}Q\\uWnO_iDCyU@",
            "sphere": 3,
          },
        ],
        "label": "O",
        "value": 538.0942,
      },
      {
        "hoses": [
          {
            "idCode": "gJQ@@dmU@\\PTFjFJ\\e}Q\\uWnO_iDCyU@",
            "sphere": 4,
          },
        ],
        "label": "O",
        "value": 538.0942,
      },
    ],
    "molfile3D": "Lattice="13.55870789 0.0 0.0 0.0 12.625570190000001 0.0 0.0 0.0 12.12849458" Properties=species:S:1:pos:R:3:GW_charged:R:1:dKS_charged:R:1:dKS_neutral:R:1 pbc="T T T"
 OpenBabel11192412253D
Lattice="13.55870789 0.0 0.0 0.0 12.625570190000001 0.0 0.0 0.0 12.12849458" Properties=species:S:1:pos:R:3:GW_charged:R:1:dKS_charged:R:1:dKS_neutral:R:1 pbc="T T T"
 15 14  0  0  0  0  0  0  0  0999 V2000
    6.5027    7.7052    6.7043 C   0  0  0  0  0  0  0  0  0  0  0  0
    6.5088    6.1739    6.6474 C   0  0  0  0  0  0  0  0  0  0  0  0
    5.0884    5.5994    6.6892 C   0  0  0  0  0  0  0  0  0  0  0  0
    7.2450    5.6603    5.4115 C   0  0  0  0  0  0  0  0  0  0  0  0
    8.6136    6.0813    5.4745 O   0  0  0  0  0  0  0  0  0  0  0  0
    7.5224    8.1070    6.6645 H   0  0  0  0  0  0  0  0  0  0  0  0
    5.9394    8.1256    5.8563 H   0  0  0  0  0  0  0  0  0  0  0  0
    6.0282    8.0633    7.6285 H   0  0  0  0  0  0  0  0  0  0  0  0
    7.0640    5.7970    7.5237 H   0  0  0  0  0  0  0  0  0  0  0  0
    4.5000    5.9303    5.8196 H   0  0  0  0  0  0  0  0  0  0  0  0
    5.0926    4.5000    6.6924 H   0  0  0  0  0  0  0  0  0  0  0  0
    4.5572    5.9340    7.5905 H   0  0  0  0  0  0  0  0  0  0  0  0
    7.1804    4.5570    5.3755 H   0  0  0  0  0  0  0  0  0  0  0  0
    6.7557    6.0561    4.5000 H   0  0  0  0  0  0  0  0  0  0  0  0
    9.0587    5.7881    4.6665 H   0  0  0  0  0  0  0  0  0  0  0  0
  1  8  1  0  0  0  0
  2  3  1  0  0  0  0
  2  1  1  0  0  0  0
  2  9  1  0  0  0  0
  3 11  1  0  0  0  0
  3 12  1  0  0  0  0
  4  5  1  0  0  0  0
  4  2  1  0  0  0  0
  6  1  1  0  0  0  0
  7  1  1  0  0  0  0
 10  3  1  0  0  0  0
 13  4  1  0  0  0  0
 14  4  1  0  0  0  0
 15  5  1  0  0  0  0
M  END
",
    "ocl": {
      "coordinates": "#qf[@Lq\\Ja[dJHjhtXOox@@Ks^hi@\`CXD~sJVAFYGCfUx\\NxCAfNiW^oPsygTA~PMlZahHQouLwJnV^lkAepk[Nh]cf\`Dgv}w~F\\k\`IiTlBwIjIoTSy|iN@@hAD\`D",
      "idCode": "gJQ@@dmU@@",
    },
    "xyz": "15
Lattice="13.55870789 0.0 0.0 0.0 12.625570190000001 0.0 0.0 0.0 12.12849458" Properties=species:S:1:pos:R:3:GW_charged:R:1:dKS_charged:R:1:dKS_neutral:R:1 pbc="T T T"
C 6.50273557 7.70524401 6.70427706
C 6.50875582 6.17394847 6.64744968
C 5.08843506 5.59936903 6.68923507
C 7.24502919 5.66032138 5.41148811
O 8.61357873 6.08129678 5.47446308
H 7.52244376 8.10703918 6.66446570
H 5.93943987 8.12557019 5.85631999
H 6.02820352 8.06329249 7.62849458
H 7.06403470 5.79698256 7.52373521
H 4.50000000 5.93029519 5.81956835
H 5.09259580 4.50000000 6.69237834
H 4.55717830 5.93395712 7.59045650
H 7.18035790 4.55699436 5.37546211
H 6.75567536 6.05610369 4.50000000
H 9.05870789 5.78808065 4.66646366",
  },
  {
    "atoms": [
      {
        "hoses": [
          {
            "idCode": "eF@Hp\\QPZQgr]W~d\`xUP",
            "sphere": 1,
          },
        ],
        "label": undefined,
        "value": 290.8475,
      },
      {
        "hoses": [
          {
            "idCode": "fH@NJ\`uOkoth\\Jh",
            "sphere": 0,
          },
        ],
        "label": "C",
        "value": 290.8475,
      },
      {
        "hoses": [
          {
            "idCode": "gCa@@duTe@xbhMDLTy|Q\\m]mRHgBj@",
            "sphere": 2,
          },
        ],
        "label": "C",
        "value": 290.8475,
      },
      {
        "hoses": [
          {
            "idCode": "gJP\`@TfzjR\`\\QTFjFJ\\e}Q\\m]mMiDSaU@",
            "sphere": 3,
          },
        ],
        "label": "C",
        "value": 290.8475,
      },
      {
        "hoses": [
          {
            "idCode": "gJP\`@TfzjR\`\\QTFjFJ\\e}Q\\m]mMiDSaU@",
            "sphere": 4,
          },
        ],
        "label": "C",
        "value": 290.8475,
      },
      {
        "hoses": [
          {
            "idCode": "fH@NJ\`uOjth\\Jh",
            "sphere": 0,
          },
        ],
        "label": "C",
        "value": 292.5165,
      },
      {
        "hoses": [
          {
            "idCode": "gCa@@duTdGDeAhabgObKek]wzQHxUP",
            "sphere": 1,
          },
        ],
        "label": "C",
        "value": 292.5165,
      },
      {
        "hoses": [
          {
            "idCode": "gJP\`@TfZjR\`\\RTFjFJ\\e}Q\\m[nMiDcaU@",
            "sphere": 2,
          },
        ],
        "label": "C",
        "value": 292.5165,
      },
      {
        "hoses": [
          {
            "idCode": "gJP\`@TfZjR\`\\RTFjFJ\\e}Q\\m[nMiDcaU@",
            "sphere": 3,
          },
        ],
        "label": "C",
        "value": 292.5165,
      },
      {
        "hoses": [
          {
            "idCode": "gJP\`@TfZjR\`\\RTFjFJ\\e}Q\\m[nMiDcaU@",
            "sphere": 4,
          },
        ],
        "label": "C",
        "value": 292.5165,
      },
      {
        "hoses": [
          {
            "idCode": "eFHBLGDTFdY|euiHOeT",
            "sphere": 1,
          },
        ],
        "label": "O",
        "value": 538.3382,
      },
      {
        "hoses": [
          {
            "idCode": "fI@GEPZgu_zTOeT",
            "sphere": 0,
          },
        ],
        "label": "O",
        "value": 538.3382,
      },
      {
        "hoses": [
          {
            "idCode": "gCa@@duTdGDEAhabgObKek]wzQ@~UP",
            "sphere": 2,
          },
        ],
        "label": "O",
        "value": 538.3382,
      },
      {
        "hoses": [
          {
            "idCode": "gJP\`@TfZjR\`\\PTFjFJ\\e}Q\\m[nMiDCyU@",
            "sphere": 3,
          },
        ],
        "label": "O",
        "value": 538.3382,
      },
      {
        "hoses": [
          {
            "idCode": "gJP\`@TfZjR\`\\PTFjFJ\\e}Q\\m[nMiDCyU@",
            "sphere": 4,
          },
        ],
        "label": "O",
        "value": 538.3382,
      },
      {
        "hoses": [
          {
            "idCode": "eMHAIhNJhMXrolnuo}IIpj\`",
            "sphere": 1,
          },
        ],
        "label": "C",
        "value": 292.1816,
      },
      {
        "hoses": [
          {
            "idCode": "fH@NJ\`uOk_th\\Jh",
            "sphere": 0,
          },
        ],
        "label": "C",
        "value": 292.1816,
      },
      {
        "hoses": [
          {
            "idCode": "gJP\`AdejjH\`\\RTFjFJ\\e}Q\\uWkO_iDcaU@",
            "sphere": 2,
          },
        ],
        "label": "C",
        "value": 292.1816,
      },
      {
        "hoses": [
          {
            "idCode": "gJP\`AdejjH\`\\RTFjFJ\\e}Q\\uWkO_iDcaU@",
            "sphere": 3,
          },
        ],
        "label": "C",
        "value": 292.1816,
      },
      {
        "hoses": [
          {
            "idCode": "gJP\`AdejjH\`\\RTFjFJ\\e}Q\\uWkO_iDcaU@",
            "sphere": 4,
          },
        ],
        "label": "C",
        "value": 292.1816,
      },
      {
        "hoses": [
          {
            "idCode": "eFHBLGDTFdY|eviHOeT",
            "sphere": 1,
          },
        ],
        "label": "O",
        "value": 538.007,
      },
      {
        "hoses": [
          {
            "idCode": "eMHAIhNHhMXrolnuo}IA|j\`",
            "sphere": 2,
          },
        ],
        "label": "O",
        "value": 538.007,
      },
      {
        "hoses": [
          {
            "idCode": "fI@GEPZgu_zTOeT",
            "sphere": 0,
          },
        ],
        "label": "O",
        "value": 538.007,
      },
      {
        "hoses": [
          {
            "idCode": "gJP\`AdejjH\`\\PTFjFJ\\e}Q\\uWkO_iDCyU@",
            "sphere": 3,
          },
        ],
        "label": "O",
        "value": 538.007,
      },
      {
        "hoses": [
          {
            "idCode": "gJP\`AdejjH\`\\PTFjFJ\\e}Q\\uWkO_iDCyU@",
            "sphere": 4,
          },
        ],
        "label": "O",
        "value": 538.007,
      },
    ],
    "molfile3D": "Lattice="12.4270524 0.0 0.0 0.0 12.614063980000001 0.0 0.0 0.0 13.01314408" Properties=species:S:1:pos:R:3:GW_charged:R:1:dKS_charged:R:1:dKS_neutral:R:1 pbc="T T T"
 OpenBabel11192412253D
Lattice="12.4270524 0.0 0.0 0.0 12.614063980000001 0.0 0.0 0.0 13.01314408" Properties=species:S:1:pos:R:3:GW_charged:R:1:dKS_charged:R:1:dKS_neutral:R:1 pbc="T T T"
 13 12  0  0  1  0  0  0  0  0999 V2000
    5.3896    7.6421    6.2970 C   0  0  0  0  0  0  0  0  0  0  0  0
    5.5053    6.1230    6.3540 C   0  0  1  0  0  0  0  0  0  0  0  0
    6.3657    5.6128    5.3136 O   0  0  0  0  0  0  0  0  0  0  0  0
    6.1164    5.6047    7.6571 C   0  0  0  0  0  0  0  0  0  0  0  0
    7.4470    6.0683    7.8591 O   0  0  0  0  0  0  0  0  0  0  0  0
    6.3661    8.1141    6.4645 H   0  0  0  0  0  0  0  0  0  0  0  0
    4.9997    7.9704    5.3219 H   0  0  0  0  0  0  0  0  0  0  0  0
    4.6940    8.0055    7.0662 H   0  0  0  0  0  0  0  0  0  0  0  0
    4.5000    5.6698    6.2527 H   0  0  0  0  0  0  0  0  0  0  0  0
    6.1866    6.1073    4.5000 H   0  0  0  0  0  0  0  0  0  0  0  0
    5.5234    5.9552    8.5131 H   0  0  0  0  0  0  0  0  0  0  0  0
    6.0806    4.5000    7.6459 H   0  0  0  0  0  0  0  0  0  0  0  0
    7.9271    5.8576    7.0400 H   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  1  6  1  0  0  0  0
  1  8  1  0  0  0  0
  2  9  1  6  0  0  0
  2  4  1  0  0  0  0
  3  2  1  0  0  0  0
  4  5  1  0  0  0  0
  4 11  1  0  0  0  0
  7  1  1  0  0  0  0
 10  3  1  0  0  0  0
 12  4  1  0  0  0  0
 13  5  1  0  0  0  0
M  END
",
    "ocl": {
      "coordinates": "#q@^q@jfhTOlwUu\\qH?Fx@@CLaN@?wFe}vEpzjZcDctmStYQDQDGKnQoolws{Hnx|AqjHFaoS|Xdm\\Fd^NiOijIaQKsaTCb\\Bcv|hj@EHAX\`D",
      "idCode": "gJP\`@TfZjR@",
    },
    "xyz": "13
Lattice="12.4270524 0.0 0.0 0.0 12.614063980000001 0.0 0.0 0.0 13.01314408" Properties=species:S:1:pos:R:3:GW_charged:R:1:dKS_charged:R:1:dKS_neutral:R:1 pbc="T T T"
C 5.38962394 7.64212192 6.29700493
C 5.50532730 6.12303379 6.35403559
O 6.36574696 5.61278149 5.31357878
C 6.11640592 5.60469055 7.65709162
O 7.44698651 6.06825426 7.85905497
H 6.36612826 8.11406398 6.46449935
H 4.99965592 7.97039841 5.32186993
H 4.69396686 8.00549178 7.06622191
H 4.50000000 5.66980810 6.25274622
H 6.18659764 6.10729219 4.50000000
H 5.52337965 5.95524325 8.51314408
H 6.08060671 4.50000000 7.64590391
H 7.92705240 5.85756557 7.03998078",
  },
  {
    "atoms": [
      {
        "hoses": [
          {
            "idCode": "eF@Hp\\QPZQgr][~d\`xUP",
            "sphere": 1,
          },
        ],
        "label": undefined,
        "value": 290.3641,
      },
      {
        "hoses": [
          {
            "idCode": "eM@Df\`xb\`ucJ~skVtdGBj@",
            "sphere": 2,
          },
        ],
        "label": "C",
        "value": 290.3641,
      },
      {
        "hoses": [
          {
            "idCode": "fH@NJ\`uOkoth\\Jh",
            "sphere": 0,
          },
        ],
        "label": "C",
        "value": 290.3641,
      },
      {
        "hoses": [
          {
            "idCode": "gJP@DejhCbB\`uPqSdojNfj}q{}H\`\\Jh",
            "sphere": 3,
          },
        ],
        "label": "C",
        "value": 290.3641,
      },
      {
        "hoses": [
          {
            "idCode": "gJP@DejhCbB\`uPqSdojNfj}q{}H\`\\Jh",
            "sphere": 4,
          },
        ],
        "label": "C",
        "value": 290.3641,
      },
      {
        "hoses": [
          {
            "idCode": "eM@HzCbJCVLk{Mk^RP\\Jh",
            "sphere": 1,
          },
        ],
        "label": "C",
        "value": 290.3779,
      },
      {
        "hoses": [
          {
            "idCode": "fH@NJ\`uOk_th\\Jh",
            "sphere": 0,
          },
        ],
        "label": "C",
        "value": 290.3779,
      },
      {
        "hoses": [
          {
            "idCode": "gJP@DijhCbB\`uPqSdojMekmq{}H\`\\Jh",
            "sphere": 2,
          },
        ],
        "label": "C",
        "value": 290.3779,
      },
      {
        "hoses": [
          {
            "idCode": "gJP@DijhCbB\`uPqSdojMekmq{}H\`\\Jh",
            "sphere": 3,
          },
        ],
        "label": "C",
        "value": 290.3779,
      },
      {
        "hoses": [
          {
            "idCode": "gJP@DijhCbB\`uPqSdojMekmq{}H\`\\Jh",
            "sphere": 4,
          },
        ],
        "label": "C",
        "value": 290.3779,
      },
      {
        "hoses": [
          {
            "idCode": "fH@NJ\`uOjth\\Jh",
            "sphere": 0,
          },
        ],
        "label": "C",
        "value": 290.5729,
      },
      {
        "hoses": [
          {
            "idCode": "gC\`@H}PGDEAhabgObKfkmwzQ@xUP",
            "sphere": 1,
          },
        ],
        "label": "C",
        "value": 290.5729,
      },
      {
        "hoses": [
          {
            "idCode": "gJP@DjZhCbB\`uPqSdojKfkmq{}H\`\\Jh",
            "sphere": 2,
          },
        ],
        "label": "C",
        "value": 290.5729,
      },
      {
        "hoses": [
          {
            "idCode": "gJP@DjZhCbB\`uPqSdojKfkmq{}H\`\\Jh",
            "sphere": 3,
          },
        ],
        "label": "C",
        "value": 290.5729,
      },
      {
        "hoses": [
          {
            "idCode": "gJP@DjZhCbB\`uPqSdojKfkmq{}H\`\\Jh",
            "sphere": 4,
          },
        ],
        "label": "C",
        "value": 290.5729,
      },
      {
        "hoses": [
          {
            "idCode": "eF@Hp\\QPZQgr]W~d\`xUP",
            "sphere": 1,
          },
        ],
        "label": "C",
        "value": 290.2787,
      },
      {
        "hoses": [
          {
            "idCode": "fH@NJ\`uOkoth\\Jh",
            "sphere": 0,
          },
        ],
        "label": "C",
        "value": 290.2787,
      },
      {
        "hoses": [
          {
            "idCode": "gC\`@DfjdPx\`hMDLTy|Qtm[nRHGBj@",
            "sphere": 2,
          },
        ],
        "label": "C",
        "value": 290.2787,
      },
      {
        "hoses": [
          {
            "idCode": "gJP@DfZjR\`\\PTFjFJ\\e}Qtm[nO_iDCaU@",
            "sphere": 3,
          },
        ],
        "label": "C",
        "value": 290.2787,
      },
      {
        "hoses": [
          {
            "idCode": "gJP@DfZjR\`\\PTFjFJ\\e}Qtm[nO_iDCaU@",
            "sphere": 4,
          },
        ],
        "label": "C",
        "value": 290.2787,
      },
      {
        "hoses": [
          {
            "idCode": "eF@Hp\\QPZQgr]W~d\`xUP",
            "sphere": 1,
          },
        ],
        "label": "C",
        "value": 290.2662,
      },
      {
        "hoses": [
          {
            "idCode": "fH@NJ\`uOkoth\\Jh",
            "sphere": 0,
          },
        ],
        "label": "C",
        "value": 290.2662,
      },
      {
        "hoses": [
          {
            "idCode": "gC\`@Dfjd\`x\`hMDLTy|Qtm[nRHGBj@",
            "sphere": 2,
          },
        ],
        "label": "C",
        "value": 290.2662,
      },
      {
        "hoses": [
          {
            "idCode": "gJP@DfZjR\`\\PTFjFJ\\e}Qtm[nO_iDCaU@",
            "sphere": 3,
          },
        ],
        "label": "C",
        "value": 290.2662,
      },
      {
        "hoses": [
          {
            "idCode": "gJP@DfZjR\`\\PTFjFJ\\e}Qtm[nO_iDCaU@",
            "sphere": 4,
          },
        ],
        "label": "C",
        "value": 290.2662,
      },
    ],
    "molfile3D": "Lattice="13.2921501 0.0 0.0 0.0 13.553625969999999 0.0 0.0 0.0 12.288030169999999" Properties=species:S:1:pos:R:3:GW_charged:R:1:dKS_charged:R:1:dKS_neutral:R:1 pbc="T T T"
 OpenBabel11192412253D
Lattice="13.2921501 0.0 0.0 0.0 13.553625969999999 0.0 0.0 0.0 12.288030169999999" Properties=species:S:1:pos:R:3:GW_charged:R:1:dKS_charged:R:1:dKS_neutral:R:1 pbc="T T T"
 17 16  0  0  0  0  0  0  0  0999 V2000
    7.7548    8.6129    5.5534 C   0  0  0  0  0  0  0  0  0  0  0  0
    7.6782    7.0847    5.5640 C   0  0  0  0  0  0  0  0  0  0  0  0
    6.2538    6.5010    5.6152 C   0  0  0  0  0  0  0  0  0  0  0  0
    6.2895    4.9782    5.4389 C   0  0  0  0  0  0  0  0  0  0  0  0
    5.5092    6.8804    6.9016 C   0  0  0  0  0  0  0  0  0  0  0  0
    8.7922    8.9559    5.4393 H   0  0  0  0  0  0  0  0  0  0  0  0
    7.1739    9.0345    4.7196 H   0  0  0  0  0  0  0  0  0  0  0  0
    7.3674    9.0536    6.4821 H   0  0  0  0  0  0  0  0  0  0  0  0
    8.1791    6.6946    4.6630 H   0  0  0  0  0  0  0  0  0  0  0  0
    8.2529    6.6910    6.4206 H   0  0  0  0  0  0  0  0  0  0  0  0
    5.6929    6.9261    4.7626 H   0  0  0  0  0  0  0  0  0  0  0  0
    6.8404    4.5000    6.2637 H   0  0  0  0  0  0  0  0  0  0  0  0
    5.2769    4.5508    5.4270 H   0  0  0  0  0  0  0  0  0  0  0  0
    6.7854    4.6934    4.5000 H   0  0  0  0  0  0  0  0  0  0  0  0
    4.5000    6.4453    6.9167 H   0  0  0  0  0  0  0  0  0  0  0  0
    6.0448    6.5057    7.7880 H   0  0  0  0  0  0  0  0  0  0  0  0
    5.3996    7.9673    7.0133 H   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  1  8  1  0  0  0  0
  2  3  1  0  0  0  0
  2 10  1  0  0  0  0
  3  5  1  0  0  0  0
  4  3  1  0  0  0  0
  4 12  1  0  0  0  0
  5 15  1  0  0  0  0
  5 17  1  0  0  0  0
  5 16  1  0  0  0  0
  6  1  1  0  0  0  0
  7  1  1  0  0  0  0
  9  2  1  0  0  0  0
 11  3  1  0  0  0  0
 13  4  1  0  0  0  0
 14  4  1  0  0  0  0
M  END
",
    "ocl": {
      "coordinates": "#qLn|x|iHaOtEQ?|MqQXH\\AhHhefD@@D\\AHHhfNa[ca@v_APH\\syWUAV{SqGTpEU]\\}MSSqW|ADYU[zTUsGmPNI]u~bvyoIBUnvsVF~XNll_ihvfqLQvnbqkFQFoKuNjncS|kJ@FhAT\`D",
      "idCode": "gJP@DjZh@",
    },
    "xyz": "17
Lattice="13.2921501 0.0 0.0 0.0 13.553625969999999 0.0 0.0 0.0 12.288030169999999" Properties=species:S:1:pos:R:3:GW_charged:R:1:dKS_charged:R:1:dKS_neutral:R:1 pbc="T T T"
C 7.75481485 8.61292208 5.55343322
C 7.67821466 7.08471020 5.56401708
C 6.25381555 6.50102015 5.61521738
C 6.28948789 4.97820523 5.43890676
C 5.50921221 6.88042037 6.90164598
H 8.79215010 8.95590863 5.43930636
H 7.17388020 9.03450902 4.71962280
H 7.36737560 9.05362597 6.48210114
H 8.17907821 6.69464952 4.66295802
H 8.25286971 6.69097634 6.42061138
H 5.69288555 6.92614764 4.76264876
H 6.84044313 4.50000000 6.26368298
H 5.27687243 4.55081102 5.42696120
H 6.78540831 4.69343960 4.50000000
H 4.50000000 6.44533746 6.91669442
H 6.04484394 6.50573198 7.78803017
H 5.39963139 7.96730808 7.01325900",
  },
]
`;
